### PR TITLE
CHECK constraint

### PR DESCRIPTION
**Pull request by nirbhayc**
.

### CODE REVIEW

File: csc2/macc_so.c [hunk 1]
```diff
@@ -137,12 +163,28 @@ void add_constraint(char *tbl, char *key)
...
 }
 
+void add_check_constraint(char *expr)
+{
+    CHECK_LEGACY_SCHEMA(1);
+    int cidx = constraints[nconstraints].ncnstrts;
+    constraints[nconstraints].ncnstrts++;
+    constraints[nconstraints].type = CONS_CHECK;
+    constraints[nconstraints].flags = 0;
+    constraints[nconstraints].ncnstrts = 1;
```
Comment (akshatsikarwar) [inline csc2/macc_so.c]: ?

File: csc2/maccparse.y [hunk 2]
```diff
@@ -146,9 +144,24 @@ cnstrtnamedstart: string T_EQ string '-' T_GT {
                       set_constraint_name($1);
                   }
 
+/* Named check constraint */
+cnstrtnamedcheck: T_CHECK string T_EQ string {
```
Comment (akshatsikarwar) [inline csc2/maccparse.y]: support double quote like index expr (anything between {} gets shipped to SQLite wholesale)

File: db/constraints.c [hunk 3]
```diff
@@ -1971,6 +1974,10 @@ int verify_constraints_exist(struct dbtable *from_db, struct dbtable *to_db,
 
     for (ii = 0; ii < from_db->n_constraints; ii++) {
         constraint_t *ct = &from_db->constraints[ii];
+
+        if (ct->type == CONS_CHECK)
```
Comment (akshatsikarwar) [inline db/constraints.c]: check !cons_fkey

File: db/record.c [hunk 4]
```diff
@@ -377,6 +377,21 @@ int add_record(struct ireq *iq, void *trans, const uint8_t *p_buf_tag_name,
         ondisktagsc = find_tag_schema(iq->usedb->tablename, ondisktag);
     }
 
+    int check_status;
+    rc = verify_check_constraints(iq->usedb, od_dta, blobs, maxblobs, 0,
```
Comment (akshatsikarwar) [inline db/record.c]: Add test case for upsert

Comment (akshatsikarwar) [reply db/record.c]: Maybe move after upsert check here..

File: db/sqlglue.c [hunk 5]
```diff
@@ -12112,7 +12156,8 @@ int indexes_expressions_data(struct schema *sc, const char *inbuf, char *outbuf,
                __func__, rc, f->name, mout.flags, f->type);
         goto done;
     }
-    if (mout.zMalloc) free(mout.zMalloc);
+    if (mout.zMalloc)
```
Comment (akshatsikarwar) [inline db/sqlglue.c]: Should failure free mout?

File: db/sqlglue.c [hunk 6]
```diff
@@ -12307,3 +12352,142 @@ struct temptable get_tbl_by_rootpg(const sqlite3 *db, int i)
...
+        rc = get_data_from_ondisk(sc, rec, blobs, maxblobs, i, &m[i], 0,
+                                  "America/New_York");
+        if (rc) {
+            logmsg(LOGMSG_ERROR, "%s: failed to get field from record",
+                   __func__);
+            goto done;
+        }
+    }
+
+    for (int i = 0; i < table->n_constraints; i++) {
```
Comment (akshatsikarwar) [inline db/sqlglue.c]: Do this work once when table is set up.

File: db/sqlglue.c [hunk 7]
```diff
@@ -12307,3 +12352,142 @@ struct temptable get_tbl_by_rootpg(const sqlite3 *db, int i)
...
+            clnt.dbglog = NULL;
+        }
+
+        /* XXX free logical tran?  */
+
+        clnt.dbtran.mode = TRANLEVEL_INVALID;
+        if (clnt.query_stats)
+            free(clnt.query_stats);
+
+        Pthread_mutex_destroy(&clnt.wait_mutex);
```
Comment (akshatsikarwar) [inline db/sqlglue.c]: Call `end_internal_sql_clnt` instead.

**Comment by akshatsikarwar** [comment]
Add testcase for check constraint on a column named 'order'

**Comment by akshatsikarwar** [comment]
Test case which exercises `escape_expr`

**Comment by akshatsikarwar** [comment]
Move out of `struct constraint`

**Comment by nirbhayc** [comment]
> Test case which exercises `escape_expr`

Not needed anymore, also have a few tests for quotes.

File: cdb2api/cdb2api.h [hunk 8]
```diff
@@ -93,6 +93,7 @@ enum cdb2_errors {
     CDB2ERR_READONLY = 305,
     CDB2ERR_TZNAME_FAIL = 401,
     CDB2ERR_CHANGENODE = 402,
+    CDB2ERR_CHECK_CONSTRAINT = 403,
```
Comment (akshatsikarwar) [inline cdb2api/cdb2api.h]: Also add to SP

Comment (akshatsikarwar) [reply cdb2api/cdb2api.h]: Test fastsql as well

**Pull request merged**